### PR TITLE
Assert can find mismatch position when using patterns

### PR DESCRIPTION
**Pull request by JanTvrdik**
Killer feature when comparing large files, such as Tracy Bluescreen outputs.

It finds the longest pattern that can be matched, then the shortest string that matches such pattern with binary search and finally it finds the longest string that matches such pattern.

### CODE REVIEW

**Comment by milo** [comment]
This is one of TODO in my head :) so :+1: Mine idea was adjust dumps to be easily comparable by some external diff. E.g. replace `%a%` pattern by spaces as long as possible.

**Comment by JanTvrdik** [comment]
I originally thought that the implementation would be a LOT simpler as I imaged just two for loops (one for $pattern and one for $actual). But `Assert::match` is so slow, that it would takes minutes to find the mismatch offset.

**Comment by JanTvrdik** [comment]
We may simplify the `*.actual` and `*.expected` files from `@outputMatch` as well but I'm not sure whether that's a good idea.

**Comment by JanTvrdik** [comment]
@milo The last commit changes the behavior to possible match what you intended to do.

**Comment by dg** [comment]
Wow

**Comment by JanTvrdik** [comment]
Now with defense against :smiling_imp: patterns! :smile:

**Comment by JanTvrdik** [comment]
@dg This is now ready to merge from my point of view.

**Comment by dg** [comment]
This is awesome!

It gives even better results when right trim (ie. `$actual = preg_replace('#^[\t ]+(?=[\r\n])#m', '', $actual);`) is applied on strings that are passed to exception. Dumper::dumpException($e) show real point of first difference.

**Comment by dg** [comment]
Btw: @milo is seems that stuff around match() is ready to be moved to a new class.

**Comment by JanTvrdik** [comment]
> It gives even better results when right trim (…) is applied on strings

I changed `expandMatchingPatterns` to return [pattern, actual] with both values normalized.

**Comment by milo** [comment]
:100: like it, I didn't forgot on this. But I would like to find time to get in deeper in it before merge.

**Comment by JanTvrdik** [comment]
@milo Last time that happen [it took months](https://github.com/nette/tester/pull/258) to get it merge.

**Comment by milo** [comment]
Yes, that's correct. Are you hurry on this?

**Comment by JanTvrdik** [comment]
I like to have as little open PR as possible. As long as this is not merged it is very difficult for me to send another non-trivial PR to Tester. There are multiple interesting problems that could be solved as a next step but I can not do it before this gets merged.

File: src/Framework/Assert.php [hunk 1]
```diff
@@ -509,6 +518,67 @@ public static function isMatching($pattern, $actual)
...
+	 * @return array
+	 * @internal
+	 */
+	public static function expandMatchingPatterns($pattern, $actual)
+	{
+		if (self::isPcre($pattern)) {
+			return [$pattern, $actual];
+		}
+
+		$parts = preg_split('#(%)#', $pattern, -1, PREG_SPLIT_DELIM_CAPTURE);
```
Comment (milo) [inline src/Framework/Assert.php]: Why is it split it by single `%` and not by `%[^%]+%`?

File: src/Framework/Assert.php [hunk 2]
```diff
@@ -509,6 +518,67 @@ public static function isMatching($pattern, $actual)
...
+		for ($i = count($parts); $i >= 0; $i--) {
+			$patternX = implode('', array_slice($parts, 0, $i));
```
Comment (milo) [inline src/Framework/Assert.php]: No need for 1st arg.

File: src/Framework/Assert.php [hunk 3]
```diff
@@ -509,6 +518,67 @@ public static function isMatching($pattern, $actual)
...
+			$patternY = "$patternX%A?%";
+			if (self::isMatching($patternY, $actual)) {
+				$patternZ = implode('', array_slice($parts, $i));
+				break;
+			}
+		}
+
+		foreach (['%A%', '%A?%'] as $evilPattern) {
```
Comment (milo) [inline src/Framework/Assert.php]: I would name it hungryPattern :)

File: src/Framework/Assert.php [hunk 4]
```diff
@@ -555,4 +625,14 @@ private static function isEqual($expected, $actual, $level = 0, $objects = NULL)
 		return $expected === $actual;
 	}
 
+
+	/**
+	 * @param  string $pattern
```
Comment (milo) [inline src/Framework/Assert.php]: No `$pattern` for consistency.

**Comment by milo** [comment]
@JanTvrdik I left few commens in diff. Maybe squash the last two commits and I'm :+1: to merge.

**Comment by milo** [comment]
@JanTvrdik May I ask for the state of this?

**Comment by JanTvrdik** [comment]
Will try to finish this today.

File: src/Framework/Assert.php (hunk 3 above)
Comment (JanTvrdik) [inline src/Framework/Assert.php]: I've renamed it o greedyPattern, is that ok?

File: src/Framework/Assert.php (hunk 1 above)
Comment (JanTvrdik) [inline src/Framework/Assert.php]: I'm not sure it would work in all cases.

**Comment by JanTvrdik** [comment]
@milo rebased and updated

**Comment by milo** [comment]
Thank you! This is really great feature.

**Pull request merged**